### ISSUE TITLE
stateful lstm fit input shape

### ISSUE THREAD

**Issue by mbinkowski** [fchollet/keras#4451]
Dear developers and users
I am working on stateful lstm and although I think I am able to train the network using `fit_generator` method, I am struggling to understand the required dimensions for the `fit` method.
Suppose I have `n_seq` training sequences of length n_time_steps and dimensionality dim. As the network is stateful, each of the batches contains several (say 10) time steps that follow the time steps from the previous batch. Therefore, I pass the argument
`batch_input_shape=(n_seq, 10, dim)`
and have my data stored in array `X` of the shape
`X.shape = (n_seq, n_time_steps, dim)`.
Now, when I use the  `fit_generator` method, with training generator yielding consecutive batches `X[:, i*10: (i + 1)*10, :]` (as presented in [keras FAQ](https://keras.io/getting-started/faq/#how-can-i-use-stateful-rnns)) the training seems to be working fine.
However, when I try to use the array `X` as an input to the `fit` method (provided that `n_time_steps % 10 == 0`) I got an error
`Error when checking model input: expected lstm_8 to have shape (4, 10, 9) but got array with shape (4, 49000, 9)`
The error seems to be raised by `keras.engine.training.standardize_input_data` function, which seems to require the input array to be of the shape passed by batch_input_shape parameter.
How to pass then more than one batch to the `fit` method?

**Comment by carlthome** [comment]
The error message should really be clarified. Anyway, what you need to do is make sure all batches are in the first dimension. E.g. reshape (4, 49000, 9) to something like (19600, 10, 9). It's kinda obvious when you think about it, but the message sure doesn't help.

**Comment by binekk** [comment]
Thanks carlthome and apologies for late reply. 

Well, I wouldn't say it's obvious that the temporal dimension need to be split between two dimensions in the input data, or that the continuation to the sequence `X[0, 0, :], X[0, 1, :], ... X[0, n_time_steps-1, :]` is `X[4, 0, :]`, especially given the [FAQ](https://keras.io/getting-started/faq/#how-can-i-use-stateful-rnns) on stateful RNNs

```
X  # this is our input data, of shape (32, 21, 16)
# we will feed it to our model in sequences of length 10

model = Sequential()
model.add(LSTM(32, batch_input_shape=(32, 10, 16), stateful=True))
model.add(Dense(16, activation='softmax'))

model.compile(optimizer='rmsprop', loss='categorical_crossentropy')

# we train the network to predict the 11th timestep given the first 10:
model.train_on_batch(X[:, :10, :], np.reshape(X[:, 10, :], (32, 16)))

# the state of the network has changed. We can feed the follow-up sequences:
model.train_on_batch(X[:, 10:20, :], np.reshape(X[:, 20, :], (32, 16)))
```
